dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/OrderAndPaginationExporterTest.java

```diff
@@ -0,0 +1,777 @@
+            .map(Event::getTrackedEntityInstance)
+            .collect(Collectors.toList());
+
+    assertEquals(List.of("QS6w44flWAf", "dUE514NMOlo"), trackedEntities);
+  }
+
+  @Test
+  void shouldSortEntitiesRespectingOrderWhenOrderParamSuppliedBeforeDataElement() {
+    EventQueryParams params = new EventQueryParams();
+    params.setOrgUnitSelectionMode(SELECTED);
+    params.setOrgUnit(orgUnit);
+    params.addDataElements(List.of(queryItem("DATAEL00006")));
+    params.addGridOrders(List.of(new OrderParam("DATAEL00006", SortDirection.DESC)));
+
+    params.addOrders(
+        List.of(
+            new OrderParam("enrolledAt", SortDirection.DESC),
+            new OrderParam("DATAEL00006", SortDirection.DESC)));
+
+    List<String> trackedEntities =
+        eventService.getEvents(params).getEvents().stream()
+            .map(Event::getTrackedEntityInstance)
+            .collect(Collectors.toList());
+
+    assertEquals(List.of("dUE514NMOlo", "QS6w44flWAf"), trackedEntities);
+  }
+
+  private static Stream<Arguments> orderByFieldInDescendingOrderWhenModeSelected() {
+    return Stream.of(
+        Arguments.of("enrollment", "pTzf9KYMk72", "D9PbzJY8bJM"),
+        Arguments.of("occurredAt", "D9PbzJY8bJM", "pTzf9KYMk72"),
+        Arguments.of("enrollmentStatus", "D9PbzJY8bJM", "pTzf9KYMk72"),
+        Arguments.of("event", "pTzf9KYMk72", "D9PbzJY8bJM"));
+  }
+
+  private static Stream<Arguments> orderByFieldInAscendingOrderWhenModeSelected() {
+    return Stream.of(
+        Arguments.of("enrollment", "D9PbzJY8bJM", "pTzf9KYMk72"),
+        Arguments.of("occurredAt", "pTzf9KYMk72", "D9PbzJY8bJM"),
+        Arguments.of("enrollmentStatus", "pTzf9KYMk72", "D9PbzJY8bJM"),
+        Arguments.of("event", "D9PbzJY8bJM", "pTzf9KYMk72"));
+  }
+
+  @ParameterizedTest
+  @MethodSource("orderByFieldInDescendingOrderWhenModeSelected")
+  void shouldOrderByFieldInDescendingOrderWhenModeSelected(
+      String field, String firstEvent, String secondEvent) {
+    EventQueryParams params = new EventQueryParams();
+    params.setOrgUnitSelectionMode(SELECTED);
+    params.setOrgUnit(orgUnit);
+    params.addOrders(List.of(new OrderParam(field, SortDirection.DESC)));
+
+    Events events = eventService.getEvents(params);
+
+    assertEquals(List.of(firstEvent, secondEvent), eventUids(events));
+  }
+
+  @ParameterizedTest
+  @MethodSource("orderByFieldInAscendingOrderWhenModeSelected")
+  void shouldOrderByFieldInAscendingOrderWhenModeSelected(
+      String field, String firstEvent, String secondEvent) {
+    EventQueryParams params = new EventQueryParams();
+    params.setOrgUnitSelectionMode(SELECTED);
+    params.setOrgUnit(orgUnit);
+    params.addOrders(List.of(new OrderParam(field, SortDirection.ASC)));
+
+    Events events = eventService.getEvents(params);
+
+    assertEquals(List.of(firstEvent, secondEvent), eventUids(events));
+  }
+
+  private static Stream<Arguments> orderByFieldInDescendingOrderWhenModeDescendants() {
```

Comment:
These don't look familiar. I don't see these on master. Should we add them on master as well?

Reply:
Sure, I can add them to master too.